mantidproject/mantid
# Additional changes for loading ESS files

**Description of work.**

Some initial changes were implemented in #28993 to allow loading ESS files.
While the algorithm returned a success status, no events were actually being read in.
This PR adds the check for the missing `offset` NX attribute (and replaces it with Unix epoch) in the remaining places.

In addition, we are now allowing **downcasting** in the `NexusIOHelper` functions through the explicit use of an optional template argument. By default, downcasting is prevented, use
```C++
auto vec = NeXus::NeXusIOHelper::readNexusSlab<
      float, NeXus::NeXusIOHelper::AllowDowncasting>(file, key, m_loadStart,
                                                     m_loadSize);
```
to allow downcasting. This is useful since some fields are 64-bits in the files, when Mantid expects 32-bits. 
No warning is logged when downcasting, I think this is ok since allowing it has to be manually enabled.
We could include the algorithm logger in the list of arguments but, this seemed to be an unnecessary complication.
**Reviewer**: is this behaviour ok? Or should the data be changed to 32-bits in the filewriter instead? Is this even possible?
It seems to me that Mantid should be a little more flexible, thus allowing downcasting, but I can see that preventing it provides more security.

I also added in the `NexusIOHelper` the possibility of reading the data into an existing buffer. The syntax is:
```C++
// no supplied buffer
auto data = readNexusVector<double>(file, "my_entry");
// with supplied buffer
std::vector<double> buffer(size);
readNexusVector(buffer, file, "my_entry");
```
Note that you need to know the size of the data to be read in advance to allocate the space in the buffer.
This can be achieved with
```C++
file.openData("my_entry");
const auto info = file.getInfo();
const auto size = std::accumulate(info.dims.begin(), info.dims.end(),
                                      int64_t{1}, std::multiplies<>());
std::vector<double> buffer(size);
```
And remember to close the file entry after reading!

**To test:**

- Generate a file using https://github.com/ess-dmsc/generate-nexus-files (e.g. the [WISH_example.py](https://github.com/ess-dmsc/generate-nexus-files/blob/master/examples/wish/WISH_example.py))
- Try to load the file with:
```Py
import mantid.simpleapi as mantid
ws = mantid.LoadEventNexus(Filename='WISH_example.nxs')
```

There was a comment by a reviewer in the previous PR (#28993 ) saying that a unit test should be included, but I think it does not serve much purpose to add a unit test for a file which may very well change in the future. I could upload the current version of the file produced by the filewriter to the Mantid ftp server and get a successful test. But I am setting up this cron-job based CI project specifically because the file generated by the filewriter is likely to change in the future, and as soon as this happens, this new unit test would be out-of-date/obsolete.

I think it is difficult to add a unit test for something that depends on multiple git repositories/projects, because they evolve independently.

*There is no associated issue.*

*This does not require release notes* because it is a very low-level minor change.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
